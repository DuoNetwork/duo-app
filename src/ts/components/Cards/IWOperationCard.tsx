import * as d3 from 'd3';
import avt from 'images/avatar.png';
import duoIcon from 'images/Duo_black.png';
import * as React from 'react';
import * as StakingCST from 'ts/common/stakingCST';
import warrentUtil from 'ts/common/warrantUtil';
import { stakeV2Wrapper, web3Wrapper } from 'ts/common/wrappers';
import {
	SCard,
	SCardTag2,
	SCardTitle,
	SStakingButtonM,
	SStakingButtonM2,
	SStakingInput
} from './_styled';

interface IProps {
	address: string;
	locale: string;
	duoBalance: number;
	award: number;
	enableApprove: boolean;
	enabled: boolean;
	phase: number;
	refresh: () => any;
}

interface IState {
	addressInfo: any;
	inputText: string;
	inputValue: number;
}

export default class IWOperationCard extends React.Component<IProps, IState> {
	constructor(props: IProps) {
		super(props);
		this.state = {
			addressInfo: {},
			inputText: '',
			inputValue: 0
		};
	}
	private handleInputChange = (value: string) => {
		const newText =
			parseInt(value, 0).toString() === 'NaN' ? '' : parseInt(value, 0).toString();
		this.setState({ inputText: newText, inputValue: parseInt(value, 0) });
	};
	private insertStake = async (txHash: string, amount: number) => {
		const item = {
			address: this.props.address,
			amount: amount.toString(),
			txHash: txHash
		};
		warrentUtil.insetStakingEntry(item);
	};

	private handleApprove = async () => {
		const { address } = this.props;
		const txHash = await web3Wrapper.erc20Approve(
			web3Wrapper.contractAddresses.DUO.address,
			address,
			web3Wrapper.contractAddresses.StakesV2[0].address,
			0,
			true
		);
		console.log('Transaction submit: ' + txHash);
	};
	private handleStake = async () => {
		const { address, locale, refresh } = this.props;
		const oracleAddr = web3Wrapper.contractAddresses.Oracles[0].address;
		const { inputValue } = this.state;
		if (inputValue >= 1) {
			const txHash = await stakeV2Wrapper.stake(address, oracleAddr, inputValue);
			this.insertStake(txHash, inputValue);
			this.setState({ inputText: '', inputValue: 0 });
			refresh();
			console.log('Transaction submit: ' + txHash);
		} else {
			window.alert(StakingCST.STK_WARING2[locale] + '1 duo');
			this.setState({ inputText: '', inputValue: 0 });
		}
	};

	private handleAutoroll = async () => {
		const { address, refresh, award } = this.props;
		const oracleAddr = web3Wrapper.contractAddresses.Oracles[0].address;
		if (award >= 1) {
			const txHash = await stakeV2Wrapper.autoRoll(address, oracleAddr, award, { gasLimit: 200000 });
			this.insertStake(txHash, award);
			this.setState({ inputText: '', inputValue: 0 });
			refresh();
		} else {
			window.alert('Reward not enough for auto roll');
		}
	};

	public render() {
		const { address, locale, duoBalance, enableApprove, enabled, award, phase } = this.props;
		const { inputText } = this.state;
		return (
			<SCard
				title={<SCardTitle>Operation</SCardTitle>}
				width="500px"
				margin="0 0 0 0"
				height="270px"
			>
				<div style={{ marginTop: 15 }}>
					<a
						style={{ color: 'rgba(0,0,0,.6)' }}
						target="_blank"
						href={
							'https://etherscan.io/token/0x56e0b2c7694e6e10391e870774daa45cf6583486?a=' +
							address
						}
					>
						<img
							style={{
								width: 16,
								height: 16,
								marginRight: 10,
								marginLeft: 5
							}}
							src={avt}
						/>
						{StakingCST.STK_ADDRESS[locale]}:{' '}
						<span style={{ color: '#5CA4DE' }}>{address}</span>
					</a>
				</div>
				<div
					style={{
						width: 455,
						display: 'flex',
						justifyContent: 'space-between'
					}}
				>
					<SCardTag2 style={{ width: 220 }}>
						<div className="bg-logo">
							<img src={duoIcon} />
						</div>
						<div className="tag-content" style={{ pointerEvents: 'none' }}>
							<div className={'tag-price USD'} style={{ fontSize: 12 }}>
								{StakingCST.STK_BALANCE[locale]}
							</div>
						</div>
						<div className="tag-subtext" style={{ pointerEvents: 'none' }}>
							<div
								style={{
									marginLeft: 20,
									fontSize: 20,
									fontWeight: 500,
									color: '#5CA4DE'
								}}
							>
								{d3.format(',.2f')(duoBalance)}
								<span style={{ fontSize: 10, marginLeft: 5 }}>DUO</span>
							</div>
						</div>
						<div
							style={{
								position: 'absolute',
								right: 10,
								top: 18
							}}
						>
							<SStakingButtonM
								onClick={this.handleApprove}
								style={{
									pointerEvents: enableApprove ? 'initial' : 'none',
									opacity: enableApprove ? 1 : 0.4
								}}
							>
								{StakingCST.STK_APPROVE[locale]}
							</SStakingButtonM>
						</div>
					</SCardTag2>
					<SCardTag2 style={{ width: 220 }}>
						<div className="bg-logo">
							<img src={duoIcon} />
						</div>
						<div className="tag-content" style={{ pointerEvents: 'none' }}>
							<div className={'tag-price USD'} style={{ fontSize: 12 }}>
								{StakingCST.STK_AWARD[locale]}
							</div>
						</div>
						<div className="tag-subtext" style={{ pointerEvents: 'none' }}>
							<div
								style={{
									marginLeft: 20,
									fontSize: 20,
									fontWeight: 500,
									color: '#5CA4DE'
								}}
							>
								{d3.format(',.2f')(award)}
							</div>
						</div>
						<div
							style={{
								position: 'absolute',
								right: 10,
								top: 18
							}}
						>
							<SStakingButtonM
								style={{
									pointerEvents: enabled ? 'initial' : 'none',
									opacity: enabled ? 1 : 0.4
								}}
								onClick={() => enabled && stakeV2Wrapper.claimReward(address)}
							>
								{StakingCST.STK_CLAIM[locale]}
							</SStakingButtonM>
						</div>
						<div
							style={{
								position: 'absolute',
								right: 10,
								top: 46
							}}
						>
							<SStakingButtonM
								onClick={() => this.handleAutoroll()}
								style={{
									pointerEvents: enabled && phase === 1 ? 'initial' : 'none',
									opacity: enabled && phase === 1 ? 1 : 0.4
								}}
							>
								{StakingCST.STK_AUTOROLL[locale]}
							</SStakingButtonM>
						</div>
					</SCardTag2>
				</div>
				<div
					style={{
						width: 455,
						marginTop: 15,
						height: 60,
						display: 'flex',
						flexDirection: 'column',
						justifyContent: 'space-between'
					}}
				>
					<div
						style={{
							width: '100%',
							height: 60,
							display: 'flex',
							flexDirection: 'column',
							justifyContent: 'space-between',
							padding: 4
						}}
					>
						<SStakingInput
							placeholder={StakingCST.STK_PLACEHODLER[locale]}
							value={inputText}
							onChange={e => this.handleInputChange(e.target.value)}
						/>
						<SStakingButtonM2
							onClick={() => this.handleStake()}
							style={{
								pointerEvents: enabled && phase === 1 ? 'initial' : 'none',
								opacity: enabled && phase === 1 ? 1 : 0.4
							}}
						>
							{StakingCST.STK_STAKE[locale]}
						</SStakingButtonM2>
					</div>
				</div>
			</SCard>
		);
	}
}
