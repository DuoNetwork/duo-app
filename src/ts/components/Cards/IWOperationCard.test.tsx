// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import { shallow } from 'enzyme';
import * as React from 'react';
import IWOperationCard from './IWOperationCard';

describe('IWOperationCard Test', () => {
	describe('Test Snapshot', () => {
		const locale = 'EN';
		const address = '0x00';
		const duoBalance = 0;
		const award = 0;
		const enableApprove = true;
		const enabled = true;
		const refresh = jest.fn(() => 123);

		it('Test Snapshot', () => {
			const wrapper = shallow(
				<IWOperationCard
					locale={locale}
					address={address}
					duoBalance={duoBalance}
					award={award}
					enableApprove={enableApprove}
					enabled={enabled}
					refresh={refresh}
				/>
			);
			expect(wrapper).toMatchSnapshot();
		});
	});
});
