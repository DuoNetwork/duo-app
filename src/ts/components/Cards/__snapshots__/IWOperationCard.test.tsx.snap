// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtendExtraDiv Test Test Snapshot Test Snapshot 1`] = `
<Styled(Card)
  height="320px"
  margin="0 0 0 0"
  title={
    <styled.div>
      Operation
    </styled.div>
  }
  width="500px"
>
  <div
    style={
      Object {
        "marginTop": 15,
      }
    }
  >
    <a
      href="https://etherscan.io/token/0x56e0b2c7694e6e10391e870774daa45cf6583486?a=0x00"
      style={
        Object {
          "color": "rgba(0,0,0,.6)",
        }
      }
      target="_blank"
    >
      <img
        src={null}
        style={
          Object {
            "height": 16,
            "marginLeft": 5,
            "marginRight": 10,
            "width": 16,
          }
        }
      />
      My Address
      :
       
      <span
        style={
          Object {
            "color": "#5CA4DE",
          }
        }
      >
        0x00
      </span>
    </a>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-between",
        "width": 455,
      }
    }
  >
    <styled.div
      style={
        Object {
          "width": 220,
        }
      }
    >
      <div
        className="bg-logo"
      >
        <img
          src={null}
        />
      </div>
      <div
        className="tag-content"
        style={
          Object {
            "pointerEvents": "none",
          }
        }
      >
        <div
          className="tag-price USD"
          style={
            Object {
              "fontSize": 12,
            }
          }
        >
          Wallet Balance
        </div>
      </div>
      <div
        className="tag-subtext"
        style={
          Object {
            "pointerEvents": "none",
          }
        }
      >
        <div
          style={
            Object {
              "color": "#5CA4DE",
              "fontSize": 20,
              "fontWeight": 500,
              "marginLeft": 20,
            }
          }
        >
          0.00
          <span
            style={
              Object {
                "fontSize": 10,
                "marginLeft": 5,
              }
            }
          >
            DUO
          </span>
        </div>
      </div>
      <div
        style={
          Object {
            "position": "absolute",
            "right": 10,
            "top": 18,
          }
        }
      >
        <styled.div
          onClick={[Function]}
          style={
            Object {
              "opacity": 1,
              "pointerEvents": "initial",
            }
          }
        >
          Approve DUO
        </styled.div>
      </div>
    </styled.div>
    <styled.div
      style={
        Object {
          "width": 220,
        }
      }
    >
      <div
        className="bg-logo"
      >
        <img
          src={null}
        />
      </div>
      <div
        className="tag-content"
        style={
          Object {
            "pointerEvents": "none",
          }
        }
      >
        <div
          className="tag-price USD"
          style={
            Object {
              "fontSize": 12,
            }
          }
        >
          Award Balance
        </div>
      </div>
      <div
        className="tag-subtext"
        style={
          Object {
            "pointerEvents": "none",
          }
        }
      >
        <div
          style={
            Object {
              "color": "#5CA4DE",
              "fontSize": 20,
              "fontWeight": 500,
              "marginLeft": 20,
            }
          }
        >
          0.00
        </div>
      </div>
      <div
        style={
          Object {
            "position": "absolute",
            "right": 10,
            "top": 18,
          }
        }
      >
        <styled.div
          onClick={[Function]}
          style={
            Object {
              "opacity": 1,
              "pointerEvents": "initial",
            }
          }
        >
          Claim Award
        </styled.div>
      </div>
    </styled.div>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "height": 60,
        "justifyContent": "space-between",
        "marginTop": 10,
        "width": 168,
      }
    }
  >
    <div
      style={
        Object {
          "border": "1px dashed rgba(0,0,0,.2)",
          "display": "flex",
          "flexDirection": "column",
          "height": 60,
          "justifyContent": "space-between",
          "padding": 2,
          "width": "100%",
        }
      }
    >
      <styled.input
        onChange={[Function]}
        placeholder="Input Stake Amount"
        value=""
      />
      <styled.div
        onClick={[Function]}
        style={
          Object {
            "opacity": 1,
            "pointerEvents": "initial",
          }
        }
      >
        Stake
      </styled.div>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "height": 60,
        "justifyContent": "space-between",
        "marginTop": 10,
        "width": 168,
      }
    }
  >
    <styled.div
      onClick={[Function]}
      style={
        Object {
          "opacity": 1,
          "pointerEvents": "initial",
        }
      }
    >
      Auto Roll
    </styled.div>
  </div>
</Styled(Card)>
`;
